From 33e4e73a17bff0823bc916998169b926115b0c6b Mon Sep 17 00:00:00 2001
From: Mike Beaton <mjsbeaton@gmail.com>
Date: Mon, 24 Nov 2025 20:40:28 +0000
Subject: [PATCH] ArmPkg: Fix errors in MiscChassisManufacturerFunction

Initial error found while attempting to implement Azure Pipelines
CLANGDWARF CI:

ERROR - Compiler #error from /__w/1/s/ArmPkg/Universal/Smbios/SmbiosMiscDx
e/Type03/MiscChassisManufacturerFunction.c variable 'ContainedElements' is
uninitialized when passed as a const pointer argument here [-Werror,-Wunin
itialized-const-pointer]

The code in question should copy the optional ContainedElements from
InputData, so fix it so that it does.

Additionally:

 - Various size calculations in the original code are incorrect, since
   the base size for calculations should not include ContainedElements,
   so fix these.
 - Since we are calculating Hdr.Length separately as part of the above,
   add code to ASSERT and return EFI_OUT_OF_RESOURCES if it would
   overflow the UINT8 location where it will be stored.
 - Since we are adding a new early exit, fix the fact that no error log
   message is generated on early exit.

Fixes: bfc0fae459543442f3f17e0de655a72aba5b0920

Signed-off-by: Mike Beaton <mjsbeaton@gmail.com>

Upstream-Status: Backport [33e4e73a17bff0823bc916998169b926115b0c6b]
Signed-off-by: Jon Mason <jon.mason@arm.com>
---
 .../Type03/MiscChassisManufacturerFunction.c  | 54 ++++++++++---------
 1 file changed, 29 insertions(+), 25 deletions(-)

diff --git a/ArmPkg/Universal/Smbios/SmbiosMiscDxe/Type03/MiscChassisManufacturerFunction.c b/ArmPkg/Universal/Smbios/SmbiosMiscDxe/Type03/MiscChassisManufacturerFunction.c
index 6b3b63b0e8d5..363f3bfe30d6 100644
--- a/ArmPkg/Universal/Smbios/SmbiosMiscDxe/Type03/MiscChassisManufacturerFunction.c
+++ b/ArmPkg/Universal/Smbios/SmbiosMiscDxe/Type03/MiscChassisManufacturerFunction.c
@@ -45,6 +45,9 @@ SMBIOS_MISC_TABLE_FUNCTION (MiscChassisManufacturer) {
   UINTN               AssertTagStrLen;
   UINTN               SerialNumStrLen;
   UINTN               ChaNumStrLen;
+  UINTN               BaseSize;
+  UINTN               ExtendLength;
+  UINTN               HdrLength;
   EFI_STRING          Manufacturer;
   EFI_STRING          Version;
   EFI_STRING          SerialNumber;
@@ -55,12 +58,6 @@ SMBIOS_MISC_TABLE_FUNCTION (MiscChassisManufacturer) {
   SMBIOS_TABLE_TYPE3  *InputData;
   EFI_STATUS          Status;
 
-  UINT8              ContainedElementCount;
-  CONTAINED_ELEMENT  ContainedElements;
-  UINT8              ExtendLength;
-
-  ExtendLength = 0;
-
   //
   // First check for invalid parameters.
   //
@@ -116,14 +113,25 @@ SMBIOS_MISC_TABLE_FUNCTION (MiscChassisManufacturer) {
   ChassisSkuNumber = HiiGetPackageString (&gEfiCallerIdGuid, TokenToGet, NULL);
   ChaNumStrLen     = StrLen (ChassisSkuNumber);
 
-  ContainedElementCount = InputData->ContainedElementCount;
-  ExtendLength          = ContainedElementCount * sizeof (CONTAINED_ELEMENT);
+  STATIC_ASSERT (OFFSET_OF (SMBIOS_TABLE_TYPE3, ContainedElements) == 0x15, "Base size of SMBIOS_TABLE_TYPE3 does not meet SMBIOS specification");
+
+  BaseSize     = OFFSET_OF (SMBIOS_TABLE_TYPE3, ContainedElements);
+  ExtendLength = (UINTN)InputData->ContainedElementCount * (UINTN)InputData->ContainedElementRecordLength;
+
+  //
+  // Length of SMBIOS struct includes ContainedElements and SKUNumber.
+  //
+  HdrLength = BaseSize + ExtendLength + sizeof (SMBIOS_TABLE_STRING);
+  if (HdrLength > MAX_UINT8) {
+    ASSERT (HdrLength <= MAX_UINT8);
+    Status = EFI_OUT_OF_RESOURCES;
+    goto Exit;
+  }
 
   //
-  // Two zeros following the last string.
+  // Additional zero follows the last string.
   //
-  RecordLength = sizeof (SMBIOS_TABLE_TYPE3) +
-                 ExtendLength    + 1 +
+  RecordLength = HdrLength       +
                  ManuStrLen      + 1 +
                  VerStrLen       + 1 +
                  SerialNumStrLen + 1 +
@@ -135,25 +143,20 @@ SMBIOS_MISC_TABLE_FUNCTION (MiscChassisManufacturer) {
     goto Exit;
   }
 
-  (VOID)CopyMem (SmbiosRecord, InputData, sizeof (SMBIOS_TABLE_TYPE3));
+  // Copy base record plus ContainedElements.
+  (VOID)CopyMem (SmbiosRecord, InputData, BaseSize + ExtendLength);
 
-  SmbiosRecord->Hdr.Length = sizeof (SMBIOS_TABLE_TYPE3) + ExtendLength + 1;
+  SmbiosRecord->Hdr.Length = HdrLength;
 
   SmbiosRecord->Type = OemGetChassisType ();
 
-  // ContainedElements
-  ASSERT (ContainedElementCount < 2);
-  (VOID)CopyMem (SmbiosRecord + 1, &ContainedElements, ExtendLength);
-
   // ChassisSkuNumber
-  SkuNumberField = (UINT8 *)SmbiosRecord +
-                   sizeof (SMBIOS_TABLE_TYPE3) -
-                   sizeof (CONTAINED_ELEMENT) + ExtendLength;
+  SkuNumberField = (UINT8 *)SmbiosRecord + BaseSize + ExtendLength;
 
+  // The string numbers in the fixed position portion of the record are populated in the input data.
   *SkuNumberField = 5;
 
-  OptionalStrStart = (CHAR8 *)((UINT8 *)SmbiosRecord + sizeof (SMBIOS_TABLE_TYPE3) +
-                               ExtendLength + 1);
+  OptionalStrStart = (CHAR8 *)((UINT8 *)SmbiosRecord + HdrLength);
   UnicodeStrToAsciiStrS (Manufacturer, OptionalStrStart, ManuStrLen + 1);
   StrStart = OptionalStrStart + ManuStrLen + 1;
   UnicodeStrToAsciiStrS (Version, StrStart, VerStrLen + 1);
@@ -175,6 +178,10 @@ SMBIOS_MISC_TABLE_FUNCTION (MiscChassisManufacturer) {
   // Now we have got the full smbios record, call smbios protocol to add this record.
   //
   Status = SmbiosMiscAddRecord ((UINT8 *)SmbiosRecord, NULL);
+
+  FreePool (SmbiosRecord);
+
+Exit:
   if (EFI_ERROR (Status)) {
     DEBUG ((
       DEBUG_ERROR,
@@ -185,9 +192,6 @@ SMBIOS_MISC_TABLE_FUNCTION (MiscChassisManufacturer) {
       ));
   }
 
-  FreePool (SmbiosRecord);
-
-Exit:
   if (Manufacturer != NULL) {
     FreePool (Manufacturer);
   }
